### PR TITLE
Mct tutorials - interface with models collection libraries

### PR DESCRIPTION
**Pull request by Idan-BenAmi**

### CODE REVIEW

File: tutorials/integrations/integrations/pytorch/helpers.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    # since we're not training, we don't need to calculate the gradients for our outputs
+    with torch.no_grad():
+        for data in data_loader:
+            images, labels = data
+            # calculate outputs by running images through the network
+            outputs = model(images.cuda())
+            # the class with the highest energy is what we choose as prediction
+            _, predicted = torch.max(outputs.data, 1)
+            total += labels.size(0)
+            correct += (predicted == labels.cuda()).sum().item()
```
Comment (eladc-git) [inline tutorials/integrations/integrations/pytorch/helpers.py:20]: labels.cuda() will crash if only cpu

File: tutorials/integrations/integrations/pytorch/timm/model_lib_timm.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+
+class ModelLib(BaseModelLib):
+
+    def __init__(self, args):
+        model_list = timm.list_models('')
+        if args[MODEL_NAME] not in model_list:
+            logging.error(f'Unknown model, Available timm models : {model_list}')
+        self.model = timm.create_model(args[MODEL_NAME], pretrained=True)
```
Comment (eladc-git) [inline tutorials/integrations/integrations/pytorch/timm/model_lib_timm.py:18]: What about the model name is a model from timm that is not classification ? this code can crash.

**Review by eladc-git — CHANGES_REQUESTED** [review]

**Review by ofirgo — CHANGES_REQUESTED** [review]
I didn't finish going over all the files but in the meantime I added some comments

**Review by ofirgo — CHANGES_REQUESTED** [review]
I didn't finish going over all the files but in the meantime I added some comments

File: tutorials/integrations/integrations/pytorch/timm/model_lib_timm.py (hunk 2 above)
Comment (Idan-BenAmi) [reply tutorials/integrations/integrations/pytorch/timm/model_lib_timm.py:18]: You are right. But I think timm is only classification... I'm trying to re-confirm

File: tutorials/integrations/main.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+                        help='The library that contains the pre-trained model')
+    parser.add_argument('--validation_dataset_folder', type=str, required=False,
+                        help='Path to the validation dataset')
+    parser.add_argument('--representative_dataset_folder', type=str, required=False,
+                        help='Path to the representative dataset used for quantization')
+    parser.add_argument('--num_representative_images', type=int, default=1024,
+                        help='Number of images for representative dataset')
+    parser.add_argument('--dataset_name', type=str, default='',
+                        help='Datset name for comment')
+    parser.add_argument('--target_platform_name', type=str, default='default',
```
[the remark this answers is not in the document]
Comment (Idan-BenAmi) [reply tutorials/integrations/main.py:25]: Agree. I'll add it to 'todo list'

File: tutorials/notebooks/example_keras_imagenet.ipynb
[the remark this answers is not in the document]
Comment (Idan-BenAmi) [reply tutorials/notebooks/example_keras_imagenet.ipynb:1]: Good idea. Added to 'todo list', I'll advise with Hai.

File: tutorials/integrations/integrations/pytorch/ultralytics/model_lib_ultralytics.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        dl = DataLoader(dataset=dataset,
+                      batch_size=batch_size,
+                      shuffle=False)
+
+        return get_representative_dataset(dl, n_iter, 'img', transforms.Normalize(0,255))
+
+    def evaluate(self, model):
+
+        # Use Cuda device if available
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (ofirgo) [inline tutorials/integrations/integrations/pytorch/ultralytics/model_lib_ultralytics.py:78]: I believe that you would need you replicate this line a lot, so maybe define this as a common utility method (this is a good example in my opinion for a function that should be in some utils file)

File: tutorials/integrations/integrations/pytorch/ultralytics/replacers.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            if isinstance(c, old_module):
+                l = new_module(get_config(c))
+                setattr(l, 'f', c.f)
+                setattr(l, 'i', c.i)
+                setattr(l, 'type', c.type)
+                setattr(m, name, l)
+    return model
+
+
+# In this section we slightly modify C2f module and replace the "list" function which is not supported by torch.fx
```
Comment (ofirgo) [inline tutorials/integrations/integrations/pytorch/ultralytics/replacers.py:33]: put all these documentation comments as a formal class documentation (also, what do you mean by "section"?)

File: tutorials/integrations/integrations/pytorch/ultralytics/replacers.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+# This missing part will be added to the postprocessing implementation
+class DetectReplacer(Detect):
+
+    def forward(self, x):
+        shape = x[0].shape  # BCHW
+        for i in range(self.nl):
+            x[i] = torch.cat((self.cv2[i](x[i]), self.cv3[i](x[i])), 1)
+        if self.training:
+            return x
+        # elif self.dynamic or self.shape != shape:
```
Comment (ofirgo) [inline tutorials/integrations/integrations/pytorch/ultralytics/replacers.py:74]: Remove commented-out lines? (here and below)

File: tutorials/integrations/main.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                        help='Specifies a target platform capabilites (tpc) name')
+    parser.add_argument('--target_platform_version', type=str, default=None,
+                        help='Specifies a target platform capabilites (tpc) version')
+    parser.add_argument('--batch_size', type=int, default=32,
+                        help='Batch size for accuracy evaluation')
+    parser.add_argument('--random_seed', type=int, default=0,
+                        help='Random seed')
```
Comment (ofirgo) [inline tutorials/integrations/main.py:32]: Improve help message (also, are you using the random seed argument to define the seed somewhere? I couldn't find it)

File: tutorials/integrations/main.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    parser.add_argument('--model_library', type=str, required=False,
+                        help='The library that contains the pre-trained model')
+    parser.add_argument('--validation_dataset_folder', type=str, required=False,
+                        help='Path to the validation dataset')
+    parser.add_argument('--representative_dataset_folder', type=str, required=False,
+                        help='Path to the representative dataset used for quantization')
+    parser.add_argument('--num_representative_images', type=int, default=1024,
+                        help='Number of images for representative dataset')
+    parser.add_argument('--dataset_name', type=str, default='',
+                        help='Datset name for comment')
```
Comment (ofirgo) [inline tutorials/integrations/main.py:24]: what d you mean "for comment"? Is it not used to choose the dataset? (if not then it's a bit confusing)

File: tutorials/integrations/main.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    parser.add_argument('--target_platform_name', type=str, default='default',
+                        help='Specifies a target platform capabilites (tpc) name')
```
Comment (ofirgo) [inline tutorials/integrations/main.py:26]: "...to choose from the provided TPCs from MCT" (or something like that)

File: tutorials/integrations/main.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    args = parser.parse_args()
+    return args
+
+
+def quantization_flow(config):
+    """
+    This function showcases the typical workflow when using MCT.
+    It begins by evaluating the performance of the floating-point model.
+    Next, the model is compressed using MCT quantization techniques.
+    Finally, the evaluation process is repeated on the compressed model to assess its performance
```
Comment (ofirgo) [inline tutorials/integrations/main.py:49]: period at the end of the sentence

**Review by ofirgo — CHANGES_REQUESTED** [review]

File: tutorials/quick_start/common/consts.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+MODEL_NAME = 'model_name'
```
Comment (ofirgo) [inline tutorials/quick_start/common/consts.py:1]: change file name to constants

File: tutorials/quick_start/pytorch_fw/timm/model_lib_timm.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    def __init__(self, args):
+        avialable_models = timm.list_models('')
+        if args[MODEL_NAME] in avialable_models:
+            self.model = timm.create_model(args[MODEL_NAME], pretrained=True)
+            self.data_config = resolve_data_config([], model=self.model)  # include the pre-processing
+            self.dataset_name = 'ImageNet'
+            super().__init__(args)
+        else:
+            raise Exception(f'Unknown model, Available timm models : {avialable_models}')
```
Comment (ofirgo) [inline tutorials/quick_start/pytorch_fw/timm/model_lib_timm.py:21]: Add model name to message: `f'Unknown model name {args[MODEL_NAME]}, available timm models are: {avialable_models}`

(and create a local variable `model_name = args[MODEL_NAME]` because you're using it several times)

File: tutorials/quick_start/pytorch_fw/timm/model_lib_timm.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+
+class ModelLib(BaseModelLib):
+
+    def __init__(self, args):
+        avialable_models = timm.list_models('')
+        if args[MODEL_NAME] in avialable_models:
+            self.model = timm.create_model(args[MODEL_NAME], pretrained=True)
+            self.data_config = resolve_data_config([], model=self.model)  # include the pre-processing
+            self.dataset_name = 'ImageNet'
```
Comment (ofirgo) [inline tutorials/quick_start/pytorch_fw/timm/model_lib_timm.py:18]: Add a comment stating that this value is constant because we currently support only timm models that are trained on Imagenet (or something like that)

File: tutorials/quick_start/pytorch_fw/timm/model_lib_timm.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+            raise Exception(f'Unknown model, Available timm models : {avialable_models}')
+
+    def get_model(self):
+        return self.model
+
+    def get_representative_dataset(self, representative_dataset_folder, n_iter, batch_size):
+        train_dataset = create_dataset(name='ImageNet', root=representative_dataset_folder, split='train',
+                                       is_training=False, batch_size=batch_size)
+        dl = create_loader(
+            train_dataset,
```
Comment (ofirgo) [inline tutorials/quick_start/pytorch_fw/timm/model_lib_timm.py:30]: Way not use constants for the strings?

File: tutorials/quick_start/pytorch_fw/torchvision/model_lib_torchvision.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        # todo: replace with dedicated API (models_list(), get_model()...) when updating to torchvision 0.14
+        return getattr(models, model_name)
+
+    @staticmethod
+    def get_torchvision_weights(model_name):
+        # todo: replace with dedicated API (models_list(), get_model()...) when updating to torchvision 0.14
+        return models.get_weight(model_name.title().replace('net', 'Net').replace('nas', 'NAS').replace('Mf', 'MF') + '_Weights.DEFAULT')
+
+    def __init__(self, args):
+        self.model = self.get_torchvision_model(args[MODEL_NAME])
```
Comment (ofirgo) [inline tutorials/quick_start/pytorch_fw/torchvision/model_lib_torchvision.py:26]: Why here we do not check that the model is available like in timm?

File: tutorials/quick_start/common/utils.py [hunk 16]
```diff
@@ -0,0 +1,31 @@
+from typing import Tuple
+import importlib
+
+
+def find_modules(lib) -> Tuple[str, str]:
```
Comment (ofirgo) [inline tutorials/quick_start/common/utils.py:5]: typehint (lib)

File: tutorials/quick_start/pytorch_fw/utils.py [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    total = 0
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    model.to(device)
+    model.eval()
+    # since we're not training, we don't need to calculate the gradients for our outputs
+    with torch.no_grad():
+        for data in data_loader:
+            images, labels = data
+            # calculate outputs by running images through the network
+            outputs = model(images.cuda())
```
Comment (ofirgo) [inline tutorials/quick_start/pytorch_fw/utils.py:17]: .cuda may fail

File: tutorials/quick_start/common/utils.py [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    if importlib.util.find_spec('pytorch_fw.' + lib) is not None:
+        model_lib_module = 'pytorch_fw.' + lib + '.model_lib_' + lib
+        quant_module = 'pytorch_fw.quant'
+
+    # Search in Keras libraries
+    elif importlib.util.find_spec('keras_fw.' + lib) is not None:
+        model_lib_module = 'keras_fw.' + lib + '.model_lib_' + lib
+        quant_module = 'keras_fw.quant'
+    else:
+        raise Exception(f'Error: model library {lib} was not found')
```
Comment (ofirgo) [inline tutorials/quick_start/common/utils.py:30]: not supported

**Review by ofirgo — CHANGES_REQUESTED** [review]

**Review by eladc-git — APPROVED** [review]
